[Logstash] Move `elastic_integration` plugin usage to ES logstash-bridge.

mashhurs opened the pull request:
## Why we are doing this?
**A bit info about plugin**
[`elastic_integration`](plugin https://github.com/elastic/logstash-filter-elastic_integration) retrieves ingest pipelines for the data stream and runs Elastic integrations it resolved. In order to do so, It internally embeds the ES sources (ingest processors, ingest plugins, etc...), will be regularly built against ES (with all active branches) and released as a [ruby gem]https://rubygems.org/gems/logstash-filter-elastic_integration

**The problem we are facing**
When ES interfaces change (for e.g, GeoIP), result in plugin build breakage. So, we in the end, came up with a solution to introduce an agreement between `elastic_integration` plugin and ES, `logstash-bridge` module! The module wraps up any classes/modules the plugin depends on and whenever interface changes, developers will apply directly to the bridge Wrapper interfaces will stay without change. This will guarantee the plugin stability.

## What this PR does?
It migrates majority plugin interface usages to the bridge. [This investigation](https://github.com/elastic/logstash-filter-elastic_integration/pull/336) shows the plugin usage details.
Exports following modules to make visible in logstash-bridge:
- `ingest-user-agent` (`org.elasticsearch.ingest.useragent`)
- `mapper-constant-keyword` (`org.elasticsearch.xpack.constantkeyword`)
- `x-pack:plugin:redact` (`org.elasticsearch.xpack.redact`)
- `:x-pack:plugin:spatial` (only required ones: `org.elasticsearch.xpack.spatial`, `org.elasticsearch.xpack.spatial.common` and `org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid`)
- `:x-pack:plugin:wildcard` (`org.elasticsearch.xpack.wildcard`)

mashhurs committed:
Testing painless processors extension.

mashhurs committed:
Test paintless spaial, wildcard and constant keyword extensions.

mashhurs committed:
Fix the bug that when null applied for Map.of() which may cause NPE.

mashhurs committed:
Debug

mashhurs committed:
Debugging: require delegate and its metadata non null to get specific exception.

mashhurs committed:
Debugging: instead of Map.copyOf(), send ingest metadata as it is.

mashhurs committed:
Make metadata version field accessible, introduce inverse wrapper with default method to avoid required interface implementation.

mashhurs committed:
RedactPlugin, IngestCommonPlugin and IngestUserAgent plugins moved to the bridge. Processor definitions also moved to the bridge. Module simplofications.

mashhurs committed:
Ingest common plugin simplification.

Commit:
[CI] Auto commit changes from spotless

mashhurs committed:
Open an access for the x-pack spatial module resources.

Commit:
[CI] Auto commit changes from spotless

mashhurs committed:
Open access to resources in mapper constant keyword.

mashhurs committed:
Add an access to resource in x-pack wildcard module.

mashhurs committed:
Export and open access to sub-package spatial modules.

mashhurs committed:
Provide module service implementations with provides keyword in x-pack spatial/wildcard/mapper-constant modules.

mashhurs committed:
Export x-pack spatial module packages to make accessible to ES server module.

mashhurs committed:
Merge branch 'main' into move-to-bridge-stable-api-investigation

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

yaauie committed:
spike refactor of logstashbridge stable API

 - transitions terminology from wrap/unwrap to toInternal/fromInternal
 - adds abstract base class for ProcessorBridge, since we are expecting external
   implementations, which includes an internal-shaped proxy to the external
   definition.
 - adds copious commentary for the classes that were previously shipped

yaauie reviewed (commented):
I think this looks on the right track -- I opened mashhurs/elasticsearch#1 to refactor things a bit, and have a comment about where the constants should live.

mashhurs committed:
Merge pull request #1 from yaauie/stable-api-clarity-external-internal

spike refactor of logstashbridge stable API

mashhurs committed:
Format the recent logstash-bridge changes.

rjernst reviewed (commented):
This bridge is getting quite a bit more complex, now trying to mimic some initialization of plugins. I want to reiterate and confirm something from when we first added it: this bridge is entirely an internal artifact of Logstash, right? In no way should this artifact ever be published for end users.

mashhurs committed:
Wildcard and mapper constant keyword modules open resource access to painless spi

mashhurs committed:
Merge branch 'main' into move-to-bridge-stable-api-investigation

mashhurs commented:
> This bridge is getting quite a bit more complex, now trying to mimic some initialization of plugins. I want to reiterate and confirm something from when we first added it: this bridge is entirely an internal artifact of Logstash, right? In no way should this artifact ever be published for end users.

👋 @rjernst, your understanding is correct that logstash-bridge is entirely internal, no end user exposure so far. From now end, https://github.com/elastic/logstash-filter-elastic_integration plugin aligns on this bridge w/o directly importing modules/classes from ES source.

> This bridge is getting quite a bit more complex, now trying to mimic some initialization of plugins. 

Historically, we got a customer failure (with below error) case which led us to initialize painless extensions (https://github.com/elastic/logstash-filter-elastic_integration/pull/162) and this is where we need to make accessible (and open an access to its resources) [redact, wildcard, mapper-constant-keyword] modules.
```
logstash-1       | Caused by: java.lang.IllegalArgumentException: cannot resolve symbol [Processors]
logstash-1       |      at org.elasticsearch.painless.PainlessScript$Script.compile(long bytes = Processors.bytes(params['size']); ctx.size_in_bytes = bytes;:24) ~[?:?]
logstash-1       |      at org.elasticsearch.painless.phase.DefaultSemanticAnalysisPhase.visitCall(DefaultSemanticAnalysisPhase.java:3233) 
```

Thank you for taking a look. We believe this modular descriptor change is beyond logstash-team scope. If you prefer to make this modular change with a separate PR (to get reviewed from corresponded team), I will make it happen but if your concern is about non-qualified `opens` statements, I have updated them to be specific (thank you for catching, I added it to spatial but somehow forgot adding to others).

rjernst commented on x-pack/plugin/mapper-constant-keyword/src/main/java/module-info.java:
```diff
@@ -0,0 +1,22 @@
+ */
+
+/** Elasticsearch X-Pack Constant Keyword Mapper Plugin. */
+module org.elasticsearch.xpack.constantkeyword {
```
you can omit the "xpack" portion of all of these module names, xpack at this point is a source code division, but they are all just ES modules.

mashhurs replied:
Ah got it (makes sense to me going w/o xpack for long term), I will!
Thank you for your prompt check.

mashhurs replied:
Addressed with [`166d4be` commit](https://github.com/elastic/elasticsearch/pull/131486/commits/166d4be3caee47d58b80cd8c9deb7bd36576f3a4)

rjernst reviewed (approved):
Thanks @mashhurs for modularizing these. The change looks fine, just one last ask on the module names.

mashhurs committed:
Rename constant-keyword, wildcard, redact and spatial modules in a way that remove xpack from namings. Move constants from processor bridge to align with a proper place where Ingest common plugin bridge is sutable.

mashhurs committed:
Merge branch 'main' into move-to-bridge-stable-api-investigation

mashhurs commented on libs/logstash-bridge/src/main/java/org/elasticsearch/logstashbridge/ingest/ProcessorBridge.java:
```diff
@@ -21,6 +23,45 @@
+        public static final String TRIM_PROCESSOR_TYPE = org.elasticsearch.ingest.common.TrimProcessor.TYPE;
+        public static final String URL_DECODE_PROCESSOR_TYPE = org.elasticsearch.ingest.common.URLDecodeProcessor.TYPE;
+        public static final String UPPERCASE_PROCESSOR_TYPE = org.elasticsearch.ingest.common.UppercaseProcessor.TYPE;
+        public static final String URI_PARTS_PROCESSOR_TYPE = org.elasticsearch.ingest.common.UriPartsProcessor.TYPE;
```
Addressed with [`166d4be` commit](https://github.com/elastic/elasticsearch/pull/131486/commits/166d4be3caee47d58b80cd8c9deb7bd36576f3a4)

mashhurs committed:
Remove unnecessary constant class in the ingest common plugin bridge.

yaauie reviewed (approved):
Fantastic work. Thank you.
